occurrences/models.py

```diff
@@ -564,7 +575,109 @@ def approved_enrolments_by_email(self, email: str):
         )
 
 
-class Enrolment(WithDeletablePersonModel):
+class EnrolmentBase(WithDeletablePersonModel):
+    notification_type = models.CharField(
+        max_length=250,
+        choices=NOTIFICATION_TYPES,
+        default=NOTIFICATION_TYPE_EMAIL,
+        verbose_name=_("notification type"),
+    )
+    enrolment_time = models.DateTimeField(
+        verbose_name=_("enrolment time"), auto_now_add=True
+    )
+    updated_at = models.DateTimeField(verbose_name=_("updated at"), auto_now=True)
+
+    class Meta:
+        abstract = True
+
+
+class EventQueueEnrolmentQuerySet(models.QuerySet):
+    def get_by_unique_id(self, unique_id):
+        compound_id = get_node_id_from_global_id(unique_id, "EventQueueEnrolmentNode")
+        enrolment_id, ts = compound_id.split("_")
+        return self.get(id=enrolment_id, enrolment_time=ts)
+
+    def with_group_occurrence_enrolment_count(self):
+        """
+        Annotate the amount of enrolments done to any occurrences
+        of the same event with the same study group name.
+
+        NOTE: The study group name is a free text field,
+        so typos in the name may affect in the result.
+        """
+        occurrence_enrolments = Enrolment.objects.filter(
+            occurrence__p_event=OuterRef("p_event"),
+            study_group__group_name=OuterRef("study_group__group_name"),
+        )
+        return self.annotate(
+            occurrence_enrolments_count=Count(
+                Subquery(occurrence_enrolments.values("pk")[:1]), distinct=True
+            )
+        )
+
+
+class EventQueueEnrolment(EnrolmentBase):
+    STATUS_HAS_NO_ENROLMENTS = "has_not_enrolled"
+    STATUS_HAS_ENROLMENTS = "has_enrolled_to_some_occurrences"
+    QUEUE_STATUSES = (
+        (
+            STATUS_HAS_NO_ENROLMENTS,
+            _("has not enrolled to any of occurrences of the event"),
+        ),
+        (
+            STATUS_HAS_ENROLMENTS,
+            _("has not enrolled to any of occurrences of the event"),
+        ),
+    )
+
+    study_group = models.ForeignKey(
+        "StudyGroup",
+        verbose_name=_("study group"),
+        related_name="queued_enrolments",
+        on_delete=models.CASCADE,
+    )
+    p_event = models.ForeignKey(
+        "PalvelutarjotinEvent",
+        verbose_name=_("event"),
+        related_name="queued_enrolments",
+        on_delete=models.CASCADE,
+    )
+
+    objects = EventQueueEnrolmentQuerySet.as_manager()
+
+    class Meta:
+        verbose_name = _("event queue enrolment")
+        verbose_name_plural = _("event queue enrolments")
+        constraints = [
+            models.UniqueConstraint(
+                fields=["study_group", "p_event"], name="unq_group_event"
```

Comment:
it's actually an old naming convention and a prefix for unique constraints. It doesn't make sense to have them named differently either, so since Khin has chosen the prefix earlier, I would like to stick with it, because it's a trivial thing after all.

Reply:
We discussed this and it was concluded that it is ok to leave this as is